component/kafka/component.go

```diff
@@ -0,0 +1,427 @@
+// The default batch size is 1 and the batch timeout is 100ms.
+// The default number of retries is 0 and the retry wait is 0.
+func New(name, group string, brokers, topics []string, proc BatchProcessorFunc, oo ...OptionFunc) (*Component, error) {
+	var errs []error
+	if name == "" {
+		errs = append(errs, errors.New("name is required"))
+	}
+
+	if group == "" {
+		errs = append(errs, errors.New("consumer group is required"))
+	}
+
+	if validation.IsStringSliceEmpty(brokers) {
+		errs = append(errs, errors.New("brokers are empty or have an empty value"))
+	}
+
+	if validation.IsStringSliceEmpty(topics) {
+		errs = append(errs, errors.New("topics are empty or have an empty value"))
+	}
+
+	if proc == nil {
+		errs = append(errs, errors.New("work processor is required"))
+	}
+
+	if len(errs) > 0 {
+		return nil, patronErrors.Aggregate(errs...)
+	}
+
+	defaultSaramaCfg, err := defaultSaramaConfig(name)
+	if err != nil {
+		return nil, err
+	}
+
+	cmp := &Component{
+		name:         name,
+		group:        group,
+		brokers:      brokers,
+		topics:       topics,
+		proc:         proc,
+		retries:      defaultRetries,
+		retryWait:    defaultRetryWait,
+		batchSize:    defaultBatchSize,
+		batchTimeout: defaultBatchTimeout,
+		failStrategy: defaultFailureStrategy,
+		saramaConfig: defaultSaramaCfg,
+	}
+
+	for _, optionFunc := range oo {
+		err = optionFunc(cmp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return cmp, nil
+}
+
+// Component is a kafka consumer implementation that processes messages in batch
+type Component struct {
+	name         string
+	group        string
+	topics       []string
+	brokers      []string
+	saramaConfig *sarama.Config
+	proc         BatchProcessorFunc
+	failStrategy FailStrategy
+	batchSize    uint
+	batchTimeout time.Duration
+	retries      uint
+	retryWait    time.Duration
+}
+
+// Run starts the consumer processing loop messages.
+func (c *Component) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	handler := newConsumerHandler(ctx, cancel, c.name, c.group, c.proc, c.failStrategy, c.batchSize,
+		c.batchTimeout, c.retries, c.retryWait)
+	client, err := sarama.NewConsumerGroup(c.brokers, c.group, c.saramaConfig)
+	if err != nil {
+		log.Errorf("error creating consumer group client for kafka consumer component: %v", err)
+		return fmt.Errorf("error creating kafka consumer component: %w", err)
+	}
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			// `Consume` should be called inside an infinite loop, when a
+			// server-side rebalance happens, the consumer session will need to be
+			// recreated to get the new claims
+			if err := client.Consume(ctx, c.topics, handler); err != nil {
+				log.Errorf("error from kafka consumer: %v", err)
+			}
+
+			// check if context was cancelled, signaling that the consumer should stop
+			if ctx.Err() != nil {
+				return
+			}
+			handler.ready = make(chan bool)
```

Comment:
It is suggested to always use `chan struct{}` because it uses no memory allocations (compiler optimization)